### DISCUSSION TITLE
Dark mode colour changes

### DISCUSSION THREAD

**Discussion by davidjgraph** [jgraph/drawio#3701] [General]
Update 2023.08.10: The new dark mode can now be turned off by unchecking Extras (or Settings), Appearance, Improve Contrast:

<img width="534" alt="Screenshot 2023-08-29 at 07 50 29" src="https://github.com/jgraph/drawio/assets/921200/d3d6093d-bf10-4ec8-8c46-63e3ac08460a">

Update 2023.07.05: These changes are out in 21.6.1. If you need to adjust colours that were originally designed just for dark mode and you now feel are wrong, use [this](https://app.diagrams.net/#_CONFIG_UzV3UjUySs1LTMpJdS4udkksyvbNT0kFiqkaO6Yl5hSnqpq7AAA=) link or set the following configuration option:

```
{
  "enableCssDarkMode": false
}
```

We've found a better system for adapting light mode designed diagrams into dark mode. _If you never use dark mode, nothing to see here for you._

This functionality relates to diagrams drawn in either mode using the default colours. These colours are changed when you switch dark/light mode. i.e. black text in light mode becomes white.

If you have a specific colour applied to a shape/text, we don't touch it when you switch modes. This leads to readability problems on the same diagram when you share it between users on different view modes.

The new functionality will store everything under the hood as the colour values in light mode and change everything, including the colour picker in dark mode. Testing has shown considerably better results over what we currently have.

However, this means that we will assume that all diagrams were drawn in light mode once the change is in place. If you drew a diagram in dark mode and pick specific colours to suit the theme of the UI, we will then take those as being the colours selected for light mode and apply the algorithm for dark mode. End result is the colours won't be the same as those you originally picked in dark mode.

This applies to the minority of users and a minority of those users' diagrams. We will provide an option in the UI to change all colours in the diagram model to the new model so you get the original colours back. That applies per diagram. 

It might seem like a heavy change, but in the long term we need the better colour adaption algorithm, the current one isn't good enough.

Note this will only apply to app.diagrams.net users and draw.io desktop users. draw.io for Confluence and Jira have never had a dark mode implement (but cloud will very soon). There maybe some downstream effect in integrations such as VSCode. We'll inform the authors that we have contact with and provide a way to switch this off if they wish.

Example diagram in light mode (original colors):
<img width="1241" alt="250078997-e7bb6c17-7b3e-47e5-8966-fee21371e34a" src="https://github.com/jgraph/drawio/assets/1761087/ade55f91-f192-46e7-b169-28fd03b712a3">

Diagram adapting to dark mode using current system:
<img width="1240" alt="250079119-15afb112-db2c-47a4-9185-ced43169d5d7" src="https://github.com/jgraph/drawio/assets/1761087/36f525ef-180d-49a2-bbe2-178ccc168372">

Using new system with same diagram:
<img width="1241" alt="250079095-2863bd47-7808-408d-a746-fc54283f159f" src="https://github.com/jgraph/drawio/assets/1761087/96383212-c19a-4ab5-be68-4c0700a1849d">

**Comment by jorgepinetta** [comment]
The dark mode looks great. Unfortunately, I must choose a light mode because I need to leave an impression.

**Comment by xeophyte** [comment]


**Reply by xknat** [reply]
@davidjgraph ^

**Reply by davidjgraph** [reply]
I didn't add those

**Reply by xknat** [reply]
Seems to be fixed :)

**Comment by vicwolk** [comment]
Damn, this is the only thing preventing me from using the dark mode! Is this available? How can I use it??

**Reply by stramin** [reply]
On the top right corner you will see a small sun icon, press there and then select "dark".

**Reply by vicwolk** [reply]
I probably misunderstood it then. I thought it would automatically convert my diagrams, which was not the case. Thank you anyways.

**Comment by xeophyte** [comment]
I hope the lines will also be visible and that the export/print options will have an option to ignore dark mode colors.

**Comment by derriche-massinissa** [comment]
Look fine and all but all my diagrams were made using dark mode, and when I opened my project this morning, the only thing I was presented with was a broken mess. This update broke most of my files. Is there a way to keep the old dark mode along this one? The old one was way better looking, this new mode looks like a cheap css filter.

**Reply by davidjgraph** [reply]
Well, it's is a cheap CSS filter... Could you please open and issue and show us an example problem?

**Reply by derriche-massinissa** [reply]
No need for an example, the images you posted are enough proof, the colors are just horrible. I like setting my colors using their hex value, so when I type FFFF00, I expect to see pure yellow and not some weird beige. This change feels like its a great thing for light mode users, but for people like me who exclusively work with dark mode (mostly because I never need to print my diagrams), this feels like a downgrade. The ideal solution would be to let us keep the old system along the new one, or at least, leave custom color untouched.

**Reply by derriche-massinissa** [reply]
And by the way, in a world where more and more tools and services are switching to dark mode as the default and prefered mode, this just feels really counter intuitive...

**Reply by davidjgraph** [reply]
I don't see anything constructive here apart from you want hex codes to match the colours you see. We can add that. Any other specific points?

**Reply by derriche-massinissa** [reply]
I think I can rephrase my issue in a better way: I prefered when the dark mode was limited to the UI and did not influence the diagram itself. So maybe separate the "dark mode" and the "invert colors" options? Or something like "improve contrast" as a checkbox and limit the css filter to the diagram instead of the whole app.

**Reply by willaCS** [reply]
I have the exact same problem, my document is made with dark ui in mind so i only use the l'element 6-15 in your picture and the ctrl-shit-6 ain't helping
Light: 
![img1](https://github.com/jgraph/drawio/assets/80853886/da5789a3-7594-44d6-9f64-f03e6bc3966f)
Dark: 
![img2](https://github.com/jgraph/drawio/assets/80853886/7fdcd394-7e94-425d-b7e6-5c6401977959)
Dark + (ctrl-shift-6)
![img3](https://github.com/jgraph/drawio/assets/80853886/e38ca667-2b99-42a4-b778-51e6c49c119c)

**Reply by alderg** [reply]
To keep using the old dark mode, use [this](https://app.diagrams.net/#_CONFIG_UzV3UjUySs1LTMpJdS4udkksyvbNT0kFiqkaO6Yl5hSnqpq7AAA=) link or set the following configuration option:

```
{
  "enableCssDarkMode": false
}
```

**Reply by derriche-massinissa** [reply]
This works like a charm! Thank you for the solution!

**Reply by willaCS** [reply]
Thank you :D
![Img4](https://github.com/jgraph/drawio/assets/80853886/2fac4163-9dcb-4cb3-927e-857b20f6f771)

**Comment by A-Ressa** [comment]
I use Background pages for Title Blocks and Borders for diagrams, and when setting a Page Background in the new version the background elements appear in Dark Mode on the pages for which they are in the background.

![bg_dark_mode_bg](https://github.com/jgraph/drawio/assets/138935856/d3539dc8-25a9-4532-af7a-6f1e2e478a9d)

![bg_dark_mode_fg](https://github.com/jgraph/drawio/assets/138935856/23e0b804-243e-4286-af50-a07e1dc3c7e9)

Full Diagram XML:
`<mxfile host="Electron" modified="2023-07-07T20:55:09.584Z" agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) draw.io/21.6.1 Chrome/112.0.5615.204 Electron/24.6.1 Safari/537.36" etag="Q-_iZxqqkgkVkyZTTWqc" version="21.6.1" type="device" pages="2">
  <diagram name="Foreground" id="snUB9yi9HBsNFyyb1NRm">
    <mxGraphModel dx="328" dy="215" grid="1" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" page="1" pageScale="1" pageWidth="1700" pageHeight="1100" backgroundImage="{&quot;src&quot;:&quot;data:page/id,drou1LnK2LpbqZTM8ZtX&quot;}" math="0" shadow="0">
      <root>
        <mxCell id="0" />
        <mxCell id="1" parent="0" />
        <mxCell id="3tSCpIjXrRqVdtPkd4sV-1" value="Foreground&lt;br&gt;Not Dark Mode" style="rounded=0;whiteSpace=wrap;html=1;" vertex="1" parent="1">
          <mxGeometry x="500" y="190" width="120" height="60" as="geometry" />
        </mxCell>
      </root>
    </mxGraphModel>
  </diagram>
  <diagram name="Background" id="drou1LnK2LpbqZTM8ZtX">
    <mxGraphModel dx="273" dy="179" grid="1" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" page="1" pageScale="1" pageWidth="1700" pageHeight="1100" math="0" shadow="0">
      <root>
        <mxCell id="OHLQMxOHJOqCZAI6xqD9-0" />
        <mxCell id="OHLQMxOHJOqCZAI6xqD9-1" parent="OHLQMxOHJOqCZAI6xqD9-0" />
        <mxCell id="vBHufXewpOdO61JFc2VS-0" value="Background&lt;br&gt;Not Dark Mode" style="ellipse;whiteSpace=wrap;html=1;labelBackgroundColor=none;" vertex="1" parent="OHLQMxOHJOqCZAI6xqD9-1">
          <mxGeometry x="340" y="180" width="120" height="80" as="geometry" />
        </mxCell>
      </root>
    </mxGraphModel>
  </diagram>
</mxfile>
`

**Comment by ghost** [comment]
would like to know if/when the cmd+shift+6 function will begin working as intended? it would be nice to use that function rather than going through config!

**Reply by davidjgraph** [reply]
This makes no difference to confluence and jira. Dark mode never existed there, so there won't be any diagrams with issues.

**Reply by ghost** [reply]
okay, then maybe I need to ask a different question. when I use #6063B2 in light mode, and then swap to dark mode, the color adapts-- which is good, as its a dark color and hard to see. but, what if I want the proper #6063B2 color in dark mode too? the color picker says it's #6063B2, but it's an adaptation to show better in dark mode, not actual #6063B2. is the only way I can achieve that color in dark mode is by doing the config?

**Reply by davidjgraph** [reply]
That needs the values in the picker corrected, you don't want to use the config for that, then you'd have a real mess. We'll take a look at the picker.

**Reply by alderg** [reply]
But even with the values in the picker corrected it's not always possible to get the exact color that was entered in dark mode. So yes, currently the only way you can achieve that color in dark mode is by doing the config, but we'll take a look at some options.

**Reply by davidjgraph** [reply]
I wouldn't use the old system on anything collaborative (i.e. jira or conf). There's 2 different problem cases here, it maybe better to discuss in a call.

**Reply by ghost** [reply]
just double checking here, is there anything you guys would like from me to help in this? I am more than happy to assist in any way I can, if that's the case. I just assumed the "we" you talk about is between you two 😅

**Reply by davidjgraph** [reply]
Yup, we're thinking about it, in between naps.

**Comment by pickVV** [comment]
Hey, I'm using draw.io in NextCloud. The color palette has recently changed. I am a dark theme user and would like to change the dark mode color palette to "normal", is there a way? I will be making more diagrams in the future and would like to have this as the default color scheme. I would also like to achieve no differences in the display of dark or light mode. (some of my colleagues use light mode, some use dark mode, and we share one big diagram that has many colors) The NextCloud version of Draw.io is 21.6.2. I tried your "Conver Colors" but most of the colors were not the same + they were pale.

Light theme:
![obrazek](https://github.com/jgraph/drawio/assets/139856027/4547047e-b45f-4035-bf16-e9546b5c5eb8)

Dark theme:
![obrazek](https://github.com/jgraph/drawio/assets/139856027/5233ac9f-a5ca-4f40-a863-73d6db80f57e)

**Comment by leiteg** [comment]
Hey @davidjgraph, thanks for this update!

One question: how are the dark colors calculated from the light colors? I had previously setup a bunch of `presetColors` to work with in dark mode, however they are all wrong now. I understand that I could turn off `enableCssDarkMode` option, but suppose I would like to use this light/dark automatic color switch. How could I setup my light preset colors such that when I activate dark mode I get my "original" preset back?

**Comment by Ndinwiddie** [comment]
Why do I have to set it to dark mode every single time I open the file?

**Comment by drhankey** [comment]
This change is rather bad UX and messes with existing data. I spent a lot of time with my team to build a diagram with a lot of color coding mostly using draw.io desktop. Obviously we were all using the dark mode. Now I wanted to distribute it to quite some people to view it with the web app and just realised that everyone will see messed up colors, unless they take extra steps. There should be a property in the data structure which indicates in which mode the diagram has been created to keep the view consistent. Or at least (as a temporary workaround) let the creator specify it manually in the file properties.

**Comment by alderg** [comment]
We've added a contrast URL parameter so you can give out the following URL https://app.diagrams.net/?dark=1&contrast=0&splash=0 or directly reference the file via it's public URL [with improved contrast](https://app.diagrams.net/?dark=1&contrast=1&url=https%3A%2F%2Fraw.githubusercontent.com%2Fjgraph%2Fdrawio-diagrams%2Fdev%2Fblog%2Finfographic-rockpool.drawio#Uhttps%3A%2F%2Fraw.githubusercontent.com%2Fjgraph%2Fdrawio-diagrams%2Fdev%2Fblog%2Finfographic-rockpool.drawio#%7B%22pageId%22%3A%22RrdJTFTAekhPfYiu2i6l%22%7D) or [without improved contrast](https://app.diagrams.net/?dark=1&contrast=0&url=https%3A%2F%2Fraw.githubusercontent.com%2Fjgraph%2Fdrawio-diagrams%2Fdev%2Fblog%2Finfographic-rockpool.drawio#Uhttps%3A%2F%2Fraw.githubusercontent.com%2Fjgraph%2Fdrawio-diagrams%2Fdev%2Fblog%2Finfographic-rockpool.drawio#%7B%22pageId%22%3A%22RrdJTFTAekhPfYiu2i6l%22%7D).

**Comment by kinglau66** [comment]
Can we have a dark mode with a light design panel? It would be great to have a button toggle to switch the appearance of the charts and grid, similar to what Outlook offers with its dark mode and light reading panel. You can find more details about this feature in Outlook here: [Dark Mode in Outlook](https://support.microsoft.com/en-us/office/dark-mode-in-outlook-3e2446e0-9a7b-4189-9af9-57fb94d02ae3).

**Comment by jxwonu** [comment]
Hello. When I use the following settings: 
![image](https://github.com/jgraph/drawio/assets/175198514/50854844-63cc-4699-aa79-5fc39062faf5)

The UI is very blurry and is very hard to read:
![image](https://github.com/jgraph/drawio/assets/175198514/d8cb9f21-ca9d-4a68-b60c-f82cbf0e6381)

How do I make it more readable while maintaining the above settings?

**Reply by davidjgraph** [reply]
That looks like a font selection problem on your system.

**Reply by jxwonu** [reply]
What do you mean? I did not change anything.

I don't have this problem in light mode. Only happens in dark mode.
![image](https://github.com/jgraph/drawio/assets/175198514/3128f307-59f3-4d46-bb30-66ce9ce78d34)

![image](https://github.com/jgraph/drawio/assets/175198514/0706e183-b4a5-476e-adcb-ce10aed998cb)

**Comment by hellt** [comment]
Hi,

I am using the embed function from the drawio, how can I prevent the css changing colors when the site where I embed those iframes toggle the preferred color theme?

I do not need to have the dark theme colors for my diagrams. The way I embed those is:

```
<div class="mxgraph" style="max-width:100%;border:1px solid transparent;margin:0 auto; display:block;" data-mxgraph='{"page":6,"zoom":2,"highlight":"#0000ff","nav":false,"check-visible-state":true,"resize":true,"url":"https://raw.githubusercontent.com/srl-labs/srlinux-vlan-handling-lab/diagrams/vlan.drawio"}'></div>

```

Here is a demo of a problem:


https://github.com/user-attachments/assets/12e31080-f210-486b-9fea-02dac8a1d7e7

PS. I would rather prefer this would have been an opt-in for the embeds, as now all my diagrams look bad when the dark theme is set in the user' system, which is a non backwards compatible change...

**Comment by alderg** [comment]
Try `"dark-mode": false` in `data-mxgraph`, ie. `<div class="mxgraph" style="max-width:100%;border:1px solid transparent;margin:0 auto; display:block;" data-mxgraph='{"dark-mode":false;"page":6,"zoom":2,"highlight":"#0000ff","nav":false,"check-visible-state":true,"resize":true,"url":"https://raw.githubusercontent.com/srl-labs/srlinux-vlan-handling-lab/diagrams/vlan.drawio"}'></div>`

**Reply by hellt** [reply]
thanks, this solves it, buuuut, I have hundreds of these across several repos...

I would understand if `"dark-mode": true` in the data-mxgraph would be an opt-in feature to not break existing integrations

**Reply by alderg** [reply]
This will be fixed in the next release. Possible values for dark-mode are `light` (default), `dark` and `auto`.
